Gansynth

jesseengel opened the pull request:
An initial demo of GANSynth in javascript. 

* Added GANSynth Model (goo.gl/magenta/gansynth-paper)
* Added short demo of generating a single note
* Added ISTFT for converting GANSynth spectral outputs back into audio.
* Memory leak still needs tracking down, but future work for @cannoneyed.

cannoneyed commented:
@jesseengel @adarob @notwaldorf anything I can do to help speed this along? Would love to get this into the repo!

notwaldorf commented:
Yikes, i didn't see that this was ready (you can add us as reviewers, then we know it's ready for a review!). Also, you have a merge conflict that needs to be resolved before the tests can run.

Looking now!

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,295 @@
+    return this.nn.predict(inputs, {batchSize}) as tf.Tensor4D;
+  }
+
+  random_sample(pitch: number) {
```
this should be camelCase

notwaldorf replied:
also needs documentation rull bad

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,295 @@
+import {MIDI_PITCHES, MIN_MIDI_PITCH, N_LATENTS, N_PITCHES} from './constants';
+import {boxUpscale, initialPad, pixelNorm} from './custom_layers';
+
+// async function dump(tensor: tf.Tensor) {
```
✂️

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    this.build(vars);
+    Object.keys(vars).map(name => vars[name].dispose());
```
i think just `vars.dispose` might work

jesseengel replied:
Unfortunately, doesn't seem to.

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,295 @@
+    });
+  }
+
+  setWeights(vars: tf.NamedTensorMap) {
```
this can probably be private

jesseengel commented on music/src/gansynth/gansynth_make_mel_sparse_coeffs.py:
```diff
@@ -0,0 +1,102 @@
+      1.0 + (frequencies_hertz / _MEL_BREAK_FREQUENCY_HERTZ))
+
+
+num_mel_bins = 1024
```
This is a python file :). I have it included because I wanted a way to recreate the coefficients file should need be.

jesseengel commented:
Okay, addressed the comments. It's not perfect yet (ex. memory leak) but @cannoneyed is going to continue working on it once we get this initial PR in.

notwaldorf commented on music/src/gansynth/custom_layers.ts:
```diff
@@ -0,0 +1,175 @@
+   * @param kwargs Only used as a pass through to call hooks.
+   * @returns A 4D `Tensor` with pixel-wise normalized channels.
+   */
+  // tslint:disable-next-line:no-any
```
(here and everywhere)

notwaldorf commented on music/src/gansynth/custom_layers.ts:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  /**
+   * @param {*} inputShapes
```
(here and everywhere)

notwaldorf commented on music/src/gansynth/custom_layers.ts:
```diff
@@ -0,0 +1,175 @@
+   */
+  // tslint:disable-next-line:no-any
+  call(inputs: tf.Tensor4D, kwargs: any): tf.Tensor4D {
+    let input = inputs;
```
i think you should wrap this in a `tf.tidy`

notwaldorf commented on music/src/gansynth/custom_layers.ts:
```diff
@@ -0,0 +1,175 @@
+   */
+  // tslint:disable-next-line:no-any
+  call(inputs: tf.Tensor4D, kwargs: any): tf.Tensor4D {
+    let input = inputs;
```
i think you should wrap this in a `tf.tidy`

notwaldorf commented on music/src/gansynth/custom_layers.ts:
```diff
@@ -0,0 +1,175 @@
+   */
+  // tslint:disable-next-line:no-any
+  call(inputs: tf.Tensor4D, kwargs: any): tf.Tensor4D {
+    let input = inputs;
```
i think you should wrap this in a `tf.tidy`

notwaldorf commented on music/src/gansynth/gansynth_make_mel_sparse_coeffs.py:
```diff
@@ -0,0 +1,102 @@
+import numpy as np
```
I'd add a huge comment at the top of this file what it is and what it generates

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,289 @@
+
+      // Stage 1
+      // tslint:disable-next-line:no-any
+      const convConfig: any = {
```
This shouldn't be `any`, make a type for it. Something like https://github.com/tensorflow/magenta-js/blob/master/music/src/core/visualizer.ts#L38

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,289 @@
+      // but have to implement with padding because python did it this way
+      // otherwise weight matrix is transposed wrong
+      const inputShape = {inputShape: [1, 1, N_LATENTS + N_PITCHES]};
+      // this.nn.add(pixelNorm(1e-8, inputShape));
```
if these aren't needed, ✂️

jesseengel replied:
I'm not 100% sure they aren't needed, because they should be there, but sound better without so I haven't debugged it totally. I'll add a TODO comment.

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,289 @@
+
+      this.outputShape = this.nn.outputShape as number[];
+      this.setWeights(vars);
+      console.log(`Output Shape:${this.outputShape}`);
```
no console.logs in the final code, please 🙏

notwaldorf commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,289 @@
+   * conditioning. Size [batch, 1, 1, N_LATENTS].
+   * @param batchSize Size of input batch.
+   */
+  predict(inputs: tf.Tensor4D, batchSize: number) {
```
add a return type,  here and to the comment

jesseengel commented on music/demos/gansynth.ts:
```diff
@@ -0,0 +1,100 @@
+
+  // GUI
+  document.getElementById('start-button').addEventListener('click', () => {
+    player.start();
```
I'm not sure what you mean? This is a `tone` player that plays the gansynth audio.

notwaldorf replied:
oh my bad, i thought it was a magenta player!

jesseengel commented on music/src/gansynth/audio_utils.ts:
```diff
@@ -0,0 +1,350 @@
+/**
+ * Utiltities for loading audio and computing mel spectrograms, based on
+ * {@link https://github.com/google/web-audio-recognition/blob/librosa-compat}.
+ * TODO(adarob): Rewrite using tfjs.
```
That's mislabelled, it was originally forked from there, but most of the code is new and specific to GANSynth. I've removed anything that was duplicate and imported it directly.

jesseengel commented on music/src/gansynth/custom_layers.ts:
```diff
@@ -0,0 +1,175 @@
+ * @param epsilon A small positive number to avoid division by zero.
+ */
+class PixelNorm extends tf.layers.Layer {
+  constructor(public epsilon = 1e-8, public layerConfig = {}) {
```
it complains if I remove them (except for the config).

teropa commented on music/src/gansynth/index.ts:
```diff
@@ -0,0 +1,18 @@
+ * =============================================================================
+ */
+
+export {GANSynth} from './model';
```
Could you also export `specgramsToAudio` so users can form the audio buffer more easily?

Like

```ts
export { specgramsToAudio } from "./audio_utils";
```

Or expose that functionality as a method in the `GANSynth` class.

jesseengel replied:
Sure thing, I'll expose it as a class.

teropa commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,292 @@
+   * conditioning. Size [batch, 1, 1, N_LATENTS].
+   * @param batchSize Size of input batch.
+   */
+  predict(inputs: tf.Tensor4D, batchSize: number) {
```
As a user I need to know the value of `N_LATENTS` as well as the pitch range of the one-hot encoding in order to form the inputs tensor. I'd prefer that to be encapsulated by the model.

`randomSample` is the right level of abstraction IMO, but for doing interpolations I need to use `predict` and these low-level details come into play.

teropa commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,292 @@
+    for (const v in vars) {
+      if (v.includes('kernel')) {
+        const fanIn = vars[v].shape[0] * vars[v].shape[1] * vars[v].shape[2];
+        vars[v] = tf.mul(vars[v], tf.sqrt(2 / fanIn));
```
I think these are leaking memory: The intermediate values of `tf.sqrt` and the overwritten value of `vars[v]` are never disposed.

teropa commented on music/src/gansynth/audio_utils.ts:
```diff
@@ -0,0 +1,231 @@
+  return istft(reIm, ispecParams);
+}
+
+export async function specgramsToAudio(specgram: tf.Tensor4D) {
```
Memory leak in this function: Needs `tidy` or disposal of tensors.

jesseengel replied:
Thanks, yah patching these up helped.

jesseengel commented:
Thanks @teropa for the valuable feedback about what would make a useful API. I've addressed the comments.

adarob commented on music/src/gansynth/audio_utils.ts:
```diff
@@ -0,0 +1,243 @@
+
+  const fftWindow = hannWindow(winLength);
+
+  // Pad the window to be the size of nFft.
```
remove here and below?

jesseengel replied:
I just removed the whole fft because I don't actually use it.

adarob commented on music/src/gansynth/model.ts:
```diff
@@ -0,0 +1,314 @@
+      this.nn.add(boxUpscale(2));
+      // Stage 2.
+      convConfig.filters = 256;
+      this.nn.add(tf.layers.conv2d(convConfig));
```
Can you add these in a for loop through the list of number of filters?

adarob commented on music/src/gansynth/audio_utils.ts:
```diff
@@ -0,0 +1,243 @@
+//@ts-ignore
+import * as FFT from 'fft.js';
+
+import {applyWindow, frame, hannWindow} from './../transcription/audio_utils';
```
please move audio_utils into core

jesseengel replied:
Maybe leave that for another time because some of these things are rather gansynth specific?

adarob commented:
I meant to move the transcription audio utils to core. We shouldn't access
one model's utils from another.

On Fri, Mar 15, 2019, 12:16 PM Jesse Engel <notifications@github.com> wrote:

> *@jesseengel* commented on this pull request.
> ------------------------------
>
> In music/src/gansynth/audio_utils.ts
> <https://github.com/tensorflow/magenta-js/pull/270#discussion_r266116320>:
>
> > + * you may not use this file except in compliance with the License.
> + * You may obtain a copy of the License at
> + *
> + * http://www.apache.org/licenses/LICENSE-2.0
> + *
> + * Unless required by applicable law or agreed to in writing, software
> + * distributed under the License is distributed on an "AS IS" BASIS,
> + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> + * See the License for the specific language governing permissions and
> + * limitations under the License.
> + */
> +import * as tf from '@tensorflow/tfjs';
> +//@ts-ignore
> +import * as FFT from 'fft.js';
> +
> +import {applyWindow, frame, hannWindow} from './../transcription/audio_utils';
>
> Maybe leave that for another time because some of these things are rather
> gansynth specific?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/magenta-js/pull/270#discussion_r266116320>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCa6PVsgMIPQp178AFE8CoeaBHk5ywRks5vW_F-gaJpZM4bLmZa>
> .
>

jesseengel commented:
Okay, addressed all the comments.
